controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,883 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"reflect"
+	"strings"
+	"time"
+
+	containersv209 "github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-09-01/containerservice"
+	containersv211 "github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-11-01/containerservice"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-10-01/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+
+	aksv1 "github.com/rancher/aks-operator/pkg/apis/aks.cattle.io/v1"
+	v10 "github.com/rancher/aks-operator/pkg/generated/controllers/aks.cattle.io/v1"
+	"github.com/rancher/aks-operator/pkg/utils"
+
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v15 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	controllerName           = "aks-controller"
+	controllerRemoveName     = "aks-controller-remove"
+	aksConfigNotCreatedPhase = ""
+	aksConfigActivePhase     = "active"
+	aksConfigUpdatingPhase   = "updating"
+	aksConfigImportingPhase  = "importing"
+	aksClusterConfigKind     = "AKSClusterConfig"
+)
+
+type Handler struct {
+	aksCC           v10.AKSClusterConfigClient
+	aksEnqueueAfter func(namespace, name string, duration time.Duration)
+	aksEnqueue      func(namespace, name string)
+	secrets         wranglerv1.SecretClient
+	secretsCache    wranglerv1.SecretCache
+}
+
+func Register(
+	ctx context.Context,
+	secrets wranglerv1.SecretController,
+	aks v10.AKSClusterConfigController) {
+
+	controller := &Handler{
+		aksCC:           aks,
+		aksEnqueue:      aks.Enqueue,
+		aksEnqueueAfter: aks.EnqueueAfter,
+		secretsCache:    secrets.Cache(),
+		secrets:         secrets,
+	}
+
+	// Register handlers
+	aks.OnChange(ctx, controllerName, controller.recordError(controller.OnAksConfigChanged))
+	aks.OnRemove(ctx, controllerRemoveName, controller.OnAksConfigRemoved)
+}
+
+func (h *Handler) OnAksConfigChanged(key string, config *aksv1.AKSClusterConfig) (*aksv1.AKSClusterConfig, error) {
+	if config == nil || config.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	ctx := context.Background()
+	switch config.Status.Phase {
+	case aksConfigImportingPhase:
+		return h.importCluster(ctx, config)
+	case aksConfigNotCreatedPhase:
+		return h.createCluster(ctx, config)
+	case aksConfigActivePhase, aksConfigUpdatingPhase:
+		return h.checkCluster(ctx, config)
+	default:
+		return config, fmt.Errorf("Invalid phase: %v", config.Status.Phase)
+	}
+	return config, nil
```

Comment:
Will this return ever be hit if the default case runs? I could be misreading this.

Reply:
Thank you Nick, even goland noticed it. I am blind